src/zenml/utils/metadata_utils.py

```diff
@@ -13,96 +13,118 @@
 #  permissions and limitations under the License.
 """Utility functions to handle metadata for ZenML entities."""
 
-import contextlib
 from typing import Dict, Optional, Union, overload
 from uuid import UUID
 
 from zenml.client import Client
-from zenml.enums import MetadataResourceTypes
+from zenml.enums import MetadataResourceTypes, ModelStages
 from zenml.logger import get_logger
 from zenml.metadata.metadata_types import MetadataType
+from zenml.models import RunMetadataResource
 from zenml.steps.step_context import get_step_context
 
 logger = get_logger(__name__)
 
 
 @overload
-def log_metadata(metadata: Dict[str, MetadataType]) -> None: ...
+def log_metadata(
+    metadata: Dict[str, MetadataType],
+) -> None: ...
 
 
 @overload
 def log_metadata(
     *,
     metadata: Dict[str, MetadataType],
-    artifact_version_id: UUID,
+    step_id: UUID,
 ) -> None: ...
 
 
 @overload
 def log_metadata(
     *,
     metadata: Dict[str, MetadataType],
-    artifact_name: str,
-    artifact_version: Optional[str] = None,
+    step_name: str,
+    run_id_name_or_prefix: Union[UUID, str],
 ) -> None: ...
 
 
 @overload
 def log_metadata(
     *,
     metadata: Dict[str, MetadataType],
-    model_version_id: UUID,
+    run_id_name_or_prefix: Union[UUID, str],
 ) -> None: ...
 
 
 @overload
 def log_metadata(
     *,
     metadata: Dict[str, MetadataType],
-    model_name: str,
-    model_version: str,
+    artifact_version_id: UUID,
 ) -> None: ...
 
 
 @overload
 def log_metadata(
     *,
     metadata: Dict[str, MetadataType],
-    step_id: UUID,
+    artifact_name: str,
+    artifact_version: Optional[str] = None,
 ) -> None: ...
 
 
 @overload
 def log_metadata(
     *,
     metadata: Dict[str, MetadataType],
-    run_id_name_or_prefix: Union[UUID, str],
+    infer_artifact: bool = False,
+    artifact_name: Optional[str] = None,
 ) -> None: ...
 
 
+# Model Metadata
 @overload
 def log_metadata(
     *,
     metadata: Dict[str, MetadataType],
-    step_name: str,
-    run_id_name_or_prefix: Union[UUID, str],
+    model_version_id: UUID,
+) -> None: ...
+
+
+@overload
+def log_metadata(
+    *,
+    metadata: Dict[str, MetadataType],
+    model_name: str,
+    model_version: Union[ModelStages, int, str],
 ) -> None: ...
 
 
+@overload
 def log_metadata(
+    *,
     metadata: Dict[str, MetadataType],
-    # Parameters to manually log metadata for steps and runs
+    infer_model: bool = False,
+) -> None: ...
+
+
+def log_metadata(
```

Comment:
I think the whole function could be a lot shorter if we just store the `RunMetadataResource.id` and `RunMetadataResource.type` and do one client call in the end?

Reply:
Makes sense. I trimmed it a bit already but kept the `resources` parameter as a list and would rather not remove it right now. I am still a bit in between if we will ever tag multiple resources at the same time and if it is alright with you, I would keep it like this.